### DISCUSSION TITLE
ESP32 freezing when makng a callback from Timers to a function that uses urequests

### DISCUSSION THREAD

**Discussion by ToyStark** [micropython/micropython#11482] [ESP32]
Hello, so here is the issue:

I have a function that send information recollected form a sensor, I need that this funcion to be called in a period bases, for this I am using Timers like this:

tim0 = Timer(0)
tim0.init(period=90000, mode=Timer.PERIODIC, callback=send_data)

I've made some test and it freezes the esp exactly in the "response = urequests.post(url, data=json_data, headers=headers)" line.

# Function to send data
def send_data(t):

    data3 = {}
    current_time = time.localtime()
    print('Enviando notificación.')
    # Restar 6 horas
    adjusted_time = list(current_time)
    adjusted_time[3] -= 6
    # Ajustar el desbordamiento de las horas (por ejemplo, si es anterior a las 6 AM)
    if adjusted_time[3] < 0:
        adjusted_time[3] += 24
    # Formatear la hora ajustada como una cadena
    time_of_catch = "{:04d}-{:02d}-{:02d} {:02d}:{:02d}:{:02d}".format(*adjusted_time)
    #print(time_of_catch)

    # La URL del servidor al que se enviará la información
    url = 'http://xxx.xxx.xxx.xx/api/notifications'

    # La información que se enviará en formato JSON
    data3 = {"message": "catchment",
            "sent_at": time_of_catch,
            "metadata": "",
            }

    # Convertir los datos a formato JSON
    json_data = json.dumps(data3)

    # Establecer la cabecera de la solicitud
    headers = {'Content-Type': 'application/json'}
    
    print('Hasta aqui es antes del urequest post')

    # Enviar la solicitud POST al servidor
    response = urequests.post(url, data=json_data, headers=headers)

    # Imprimir la respuesta del servidor
    print(response.text)

If i call the function directly from main, as shown below, it works perfectly fine, it seems that the combination of callbacks and urequest have some issues.

#Function main
def main():

    wifiConncect()
    send_data()

Any assistance will be very much appriciated :)

**Comment by karfas** [comment]
See
- https://docs.micropython.org/en/latest/library/machine.Timer.html?highlight=timer#machine.Timer
- https://docs.micropython.org/en/latest/reference/isr_rules.html#isr-rules

Why do we dare to create documentation ? Nobody seems to read it.

**Reply by ToyStark** [reply]
Thank you, I'm new in this, I'll take a look in this. Thnks again.

**Comment by bidrohini** [comment]
The freezing issue you are experiencing when using the urequests library inside a callback function is likely due to the blocking nature of the library. The urequests.post() function blocks the execution until it receives a response from the server, which can cause the ESP to freeze or become unresponsive.

To avoid this issue, you can try using the non-blocking version of HTTP requests, such as the uasyncio library, which provides asynchronous capabilities to MicroPython. Here's an example of how you can modify your code to use uasyncio:
import uasyncio as asyncio
import urequests
import json

# Function to send data
async def send_data():
    data3 = {}
    current_time = time.localtime()
    print('Enviando notificación.')
    # Restar 6 horas
    adjusted_time = list(current_time)
    adjusted_time[3] -= 6
    # Ajustar el desbordamiento de las horas (por ejemplo, si es anterior a las 6 AM)
    if adjusted_time[3] < 0:
        adjusted_time[3] += 24
    # Formatear la hora ajustada como una cadena
    time_of_catch = "{:04d}-{:02d}-{:02d} {:02d}:{:02d}:{:02d}".format(*adjusted_time)
    # print(time_of_catch)

    # La URL del servidor al que se enviará la información
    url = 'http://xxx.xxx.xxx.xx/api/notifications'

    # La información que se enviará en formato JSON
    data3 = {"message": "catchment",
             "sent_at": time_of_catch,
             "metadata": "",
             }

    # Convertir los datos a formato JSON
    json_data = json.dumps(data3)

    # Establecer la cabecera de la solicitud
    headers = {'Content-Type': 'application/json'}

    print('Hasta aqui es antes del urequest post')

    # Enviar la solicitud POST al servidor
    response = urequests.post(url, data=json_data, headers=headers)

    # Imprimir la respuesta del servidor
    print(response.text)

async def main():
    wifiConnect()
    while True:
        await send_data()
        await asyncio.sleep(90)

# Run the event loop
loop = asyncio.get_event_loop()
loop.run_until_complete(main())


   
In this modified version, we use uasyncio to create an event loop and schedule the send_data() function to be called periodically using asyncio.sleep(). The send_data() function is defined as an asynchronous function, allowing it to run concurrently with other tasks.

Note that you will need to have the uasyncio and urequests libraries installed on your ESP board. Make sure to install them before running the modified code.

By using uasyncio, you can achieve non-blocking behavior and prevent the freezing issue caused by blocking operations in a callback function.
